bench/irmin-pack/tree.ml

```diff
@@ -49,7 +35,148 @@ let pp_store_type ppf = function
+    if hl < 0. then invalid_arg "Wrong half life";
+    (if hl = 0. then 0. else log 0.5 /. hl |> exp) |> create
+
+  (** [from_half_life_ratio hl_ratio step_count] is [ema], a functional
+      exponential moving average. After [hl_ratio * step_count] calls to
+      [update], half of the past is forgotten. *)
+  let from_half_life_ratio hl_ratio step_count =
+    if hl_ratio < 0. then invalid_arg "Wrong half life ratio";
+    if step_count <= 0L then invalid_arg "Wront step count";
+    Int64.to_float step_count *. hl_ratio |> from_half_life
+
+  (** Feed a new point to the EMA. *)
+  let update ema point =
+    let value = (ema.value *. ema.momentum) +. (point *. ema.opp_momentum) in
+    let count = ema.count +. 1. in
+    { ema with value; count }
+
+  (** Read the EMA value. *)
+  let peek ema =
+    if ema.count = 0. then failwith "Can't peek an EMA before first update";
+    ema.value /. (1. -. (ema.momentum ** ema.count))
+end
+
+module Bootstrap_trace = struct
+  type 'a scope = Forget of 'a | Keep of 'a [@@deriving repr]
+  type key = string list [@@deriving repr]
+  type hash = string [@@deriving repr]
+  type message = string [@@deriving repr]
+
+  type add = {
+    key : key;
+    value : string;
+    ctx0 : int64 scope;
+    ctx1 : int64 scope;
+  }
+  [@@deriving repr]
+
+  type copy = {
+    key_src : key;
+    key_dst : key;
+    ctx0 : int64 scope;
+    ctx1 : int64 scope;
+  }
+  [@@deriving repr]
+
+  type commit = {
+    hash : hash scope;
+    date : int64;
+    message : message;
+    parents : hash scope list;
+    ctx0 : int64 scope;
+  }
+  [@@deriving repr]
+
+  (** The 8 different operations recorded in Tezos.
+
+      {3 Interleaved Contexts and Commits}
+
+      All the recorded operations in Tezos operate on (and create new) immutable
+      records of type [context]. Most of the time, everything is linear (i.e.
+      the input context to an operation is the latest output context), but there
+      sometimes are several parallel chains of contexts, where all but one will
+      end up being discarded.
+
+      Similarly to contexts, commits are not always linear, i.e. a checkout may
+      choose a parent that is not the latest commit.
+
+      To solve this conundrum when replaying the trace, we need to remember all
+      the [context_id -> tree] and [trace commit hash -> real commit hash] pairs
+      to make sure an operation is operating on the right parent.
+
+      In the trace, the context indices and the commit hashes are 'scoped`,
+      meaning that they are tagged with a boolean information indicating if this
+      is the very last occurence of that value in the trace. This way we can
+      discard a remembered pair as soon as possible.
+
+      In practice, there is only 1 context and 1 commit in history, and
+      sometimes 0 or 2, but the code is ready for more. *)
+  type op =
+    (* Operation(s) that create a context from none *)
+    | Checkout of hash scope * int64 scope
```

Comment:
can you introduce a type alias for the context id too?

Reply:
👍